File: old/clara/sim_rwa.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8
#
# GA: RWA with GOF
# Genetic Algorithm: 
# Routing and Wavelength Assignment with General Objective Function
#
# Copyright 2017 Universidade Federal do Pará (PPGCC UFPA)
#
# Authors: April 2017
# Cassio Trindade Batista - cassio.batista.13@gmail.com

# Last revised on Apr 2017

# REFERENCES:
# [1] 
# Afonso Jorge F. Cardoso et. al., 2010
# A New Proposal of an Efficient Algorithm for Routing and Wavelength 
# Assignment (RWA) in Optical Networks

import sys

import info
import nsf

import numpy as np

from rwa_ga import rwa_ga

from rwa_std_fix import rwa_std_fix
from rwa_std_alt import rwa_std_alt

from rwa_alt import rwa_alt
from rwa_fix import rwa_fix

def get_wave_availability(k, n):
	return (int(n) & ( 1 << k )) >> k

if __name__ == '__main__':

	nsf_wave, nsf_adj, nsf_time, nsf_links = nsf.generate()

	blocked_ga  = []
	blocked_std_fix = []
	blocked_std_alt = []
	blocked_alt = []
	blocked_fix = []

	if info.DEBUG:
		print nsf_wave

	for load in xrange(info.SIM_MIN_LOAD, info.SIM_MAX_LOAD):
		N_ga  = nsf_wave.copy()
		N_std_fix = nsf_wave.copy()
		N_std_alt = nsf_wave.copy()
		N_alt = nsf_wave.copy()
		N_fix = nsf_wave.copy()

		T_ga  = nsf_time.copy() # holding time
		T_std_fix = nsf_time.copy() # holding time
		T_std_alt = nsf_time.copy() # holding time
		T_alt = nsf_time.copy() # holding time
		T_fix = nsf_time.copy() # holding time
	
		paths_fix = []
		paths_alt = []

		count_block_ga  = 0
		count_block_std_fix = 0
		count_block_std_alt = 0
		count_block_alt = 0
		count_block_fix = 0

		for gen in xrange(info.SIM_NUM_GEN):
			until_next   = -np.log(1-np.random.rand())/load
			holding_time = -np.log(1-np.random.rand())
	
			count_block_ga  += rwa_ga(N_ga,  nsf_adj, T_ga,  holding_time)
			count_block_std_fix += rwa_std_fix(N_std_fix, nsf_adj, T_std_fix, holding_time, paths_fix)
			count_block_std_alt += rwa_std_alt(N_std_alt, nsf_adj, T_std_alt, holding_time, paths_alt)
			count_block_alt += rwa_alt(N_alt, nsf_adj, T_alt, holding_time)
			count_block_fix += rwa_fix(N_fix, nsf_adj, T_fix, holding_time)

			if info.DEBUG:
				sys.stdout.write('\rLoad: %02d/%02d ' % (load, info.SIM_MAX_LOAD-1))
				sys.stdout.write('Simul: %04d/%04d\t' % (gen+1, info.SIM_NUM_GEN))
				sys.stdout.write('GA:  %04d, ' % count_block_ga)
				sys.stdout.write('STF: %04d, ' % count_block_std_fix)
				sys.stdout.write('STA: %04d, ' % count_block_std_alt)
				sys.stdout.write('ALT: %04d, ' % count_block_alt)
				sys.stdout.write('FIX: %04d '  % count_block_fix)
				sys.stdout.flush()
	
			aux_fix = {idx:[] for idx in xrange(info.NSF_NUM_CHANNELS)}
			aux_alt = {idx:[] for idx in xrange(info.NSF_NUM_CHANNELS)}
			# Atualiza os todos os canais que ainda estao sendo usados 
			for link in nsf_links:
				i, j = link
				for w in xrange(info.NSF_NUM_CHANNELS):

					# GA + GOF
					if  T_ga[i][j][w] > until_next:
						T_ga[i][j][w] -= until_next
						T_ga[j][i][w]  = T_ga[i][j][w]
					else:
						T_ga[i][j][w] = 0
						T_ga[j][i][w] = 0
						if not get_wave_availability(w, N_ga[i][j]):
							N_ga[i][j] += 2**w # free channel
							N_ga[j][i] = N_ga[i][j] 

					# Dijkstra + Graph coloring
					if  T_std_fix[i][j][w] > until_next:
						T_std_fix[i][j][w] -= until_next
						T_std_fix[j][i][w]  = T_std_fix[i][j][w]
					else:
						T_std_fix[i][j][w] = 0
						T_std_fix[j][i][w] = 0
						if not get_wave_availability(w, N_std_fix[i][j]):
							N_std_fix[i][j] += 2**w # free channel
							N_std_fix[j][i] = N_std_fix[i][j] 
							aux_fix[w].append((i,j))
							aux_fix[w].append((j,i))

					# Yen + Graph coloring
					if  T_std_alt[i][j][w] > until_next:
						T_std_alt[i][j][w] -= until_next
						T_std_alt[j][i][w]  = T_std_alt[i][j][w]
					else:
						T_std_alt[i][j][w] = 0
						T_std_alt[j][i][w] = 0
						if not get_wave_availability(w, N_std_alt[i][j]):
							N_std_alt[i][j] += 2**w # free channel
							N_std_alt[j][i] = N_std_alt[i][j] 
							aux_alt[w].append((i,j))
							aux_alt[w].append((j,i))

					# Yen + First-fit
					if  T_alt[i][j][w] > until_next:
						T_alt[i][j][w] -= until_next
						T_alt[j][i][w]  = T_alt[i][j][w]
					else:
						T_alt[i][j][w] = 0
						T_alt[j][i][w] = 0
						if not get_wave_availability(w, N_alt[i][j]):
							N_alt[i][j] += 2**w # free channel
							N_alt[j][i] = N_alt[i][j] 

					# Dijkstra + First-fit
					if  T_fix[i][j][w] > until_next:
						T_fix[i][j][w] -= until_next
						T_fix[j][i][w]  = T_fix[i][j][w]
					else:
						T_fix[i][j][w] = 0
						T_fix[j][i][w] = 0
						if not get_wave_availability(w, N_fix[i][j]):
							N_fix[i][j] += 2**w # free channel
							N_fix[j][i] = N_fix[i][j] 

			pop_fix = []
			for path in paths_fix:
				R, color = path
				count = 0
				if color == None:
					continue
				for r in xrange(len(R)-1):
					rcurr = R[r]
					rnext = R[r+1]
					if (rcurr,rnext) in aux_fix[color]:
						count += 1
				if count == len(R)-1:
					pop_fix.append(paths_fix.index(path))

			pop_fix.sort() # make sure the last elements are popped first
			while len(pop_fix):
				paths_fix.pop(pop_fix.pop())

			pop_alt = []
			for path in paths_alt:
				R, color = path
				count = 0
				if color == None:
					continue
				for r in xrange(len(R)-1):
					rcurr = R[r]
					rnext = R[r+1]
					if (rcurr,rnext) in aux_alt[color]:
						count += 1
				if count == len(R)-1:
					pop_alt.append(paths_alt.index(path))

			pop_alt.sort() # make sure the last elements are popped first
			while len(pop_alt):
				paths_alt.pop(pop_alt.pop())

		blocked_ga.append( 100.0*count_block_ga /info.SIM_NUM_GEN)
		blocked_std_fix.append(100.0*count_block_std_fix/info.SIM_NUM_GEN)
		blocked_std_alt.append(100.0*count_block_std_alt/info.SIM_NUM_GEN)
		blocked_alt.append(100.0*count_block_alt/info.SIM_NUM_GEN)
		blocked_fix.append(100.0*count_block_fix/info.SIM_NUM_GEN)
		print 'Done'

	if info.DEBUG:
		print '\tGA:  ', blocked_ga
		print '\tSTF: ', blocked_std_fix
		print '\tSTA: ', blocked_std_alt
		print '\tALT: ', blocked_alt
		print '\tFIX: ', blocked_fix
	
	with open('block_GA_%d.m' % info.NSF_NUM_CHANNELS, 'a') as f:
		text = str(blocked_ga).replace('[','').replace(']','')
		f.write('%s; ...\n' % text)
	
	with open('block_STF_%d.m' % info.NSF_NUM_CHANNELS, 'a') as f:
		text = str(blocked_std_fix).replace('[','').replace(']','')
		f.write('%s; ...\n' % text)
	
	with open('block_STA_%d.m' % info.NSF_NUM_CHANNELS, 'a') as f:
		text = str(blocked_std_alt).replace('[','').replace(']','')
		f.write('%s; ...\n' % text)
	
	with open('block_ALT_%d.m' % info.NSF_NUM_CHANNELS, 'a') as f:
		text = str(blocked_alt).replace('[','').replace(']','')
		f.write('%s; ...\n' % text)
	
	with open('block_FIX_%d.m' % info.NSF_NUM_CHANNELS, 'a') as f:
		text = str(blocked_fix).replace('[','').replace(']','')
		f.write('%s; ...\n' % text)

### EOF ###

File: c/sim_rwa.c
/*
 * GA: RWA with GOF
 * Genetic Algorithm  
 * Routing and Wavelength Assignment
 * General Objective Function
 *
 * Copyright 2017
 * Programa de Pós-Graduação em Ciência da Computação (PPGCC)
 * Universidade Federal do Pará (UFPA)
 *
 * Author: July 2017
 * Cassio Trindade Batista - cassio.batista.13@gmail.com
 *
 * Last edited on July, 2017
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <math.h>

#include "info.h"
#include "list.h"
#include "nsf.h"
#include "rwa_ga.h"

// https://stackoverflow.com/questions/2438142/dynamic-memory-allocation-for-3d-array
// https://stackoverflow.com/questions/2128728/allocate-matrix-in-c
// https://stackoverflow.com/questions/1694827/random-float-number
// https://stackoverflow.com/questions/7109964/creating-your-own-header-file-in-c

int
main(int argc, char** argv)
{
	int i, j, w; 
	int load, gen;
	
	float until_next, holding_time;
	int count_block;

	nsf* nsfpack;
	nsf* nsf_net;

	FILE* fp = NULL;
	if(!(fp = fopen("block.txt", "w"))) {
		printf("Failed to open block file\n");
		exit(0);
	}

	srand(time(NULL));

	/* init NSF network data structure */
	nsfpack = (nsf*)malloc(sizeof(nsf));
	nsfpack = nsf_generate(nsfpack);

	for(load=SIM_MIN_LOAD; load<SIM_MAX_LOAD; load++) {
		nsf_net = (nsf*)malloc(sizeof(nsf));
		nsf_net = nsf_deep_copy(nsfpack, nsf_net);

		count_block = 0;

		for(gen=0; gen<SIM_NUM_GEN; gen++) {
			until_next   = -logf(1-(rand()/(float)RAND_MAX))/load;
			holding_time = -logf(1-(rand()/(float)RAND_MAX));

			printf("before:\n");
			for(j=0; j<nsf_net->j; j++) {
				for(w=0; w<nsf_net->w; w++) {
					printf("%d ", nsf_net->N[0][j][w]);
				}
				printf("\n");
			}

			count_block += rwa(&nsf_net, holding_time);

			printf("after:\n");
			for(j=0; j<nsf_net->j; j++) {
				for(w=0; w<nsf_net->w; w++) {
					printf("%d ", nsf_net->N[0][j][w]);
				}
				printf("\n");
			}

			if(!DEBUG) {
				printf("\rLoad: %02d/%02d ", load, SIM_MAX_LOAD-1);
				printf("Simul: %04d/%04d\t", gen+1, SIM_NUM_GEN);
				printf("GA:  %04d, ", count_block);
			}

			// Atualiza os todos os canais que ainda estao sendo usados */
			for(i=0; i<NSF_NUM_NODES; i++) {
				for(j=i+1; j<NSF_NUM_NODES; j++) {
					if(!nsf_net->A[i][j]) // if there's no adjacency, try another node
						continue;

					for(w=0; w<NSF_NUM_CHANNELS; w++) {
						if(nsf_net->T[i][j][w] > until_next) {
							nsf_net->T[i][j][w] = nsf_net->T[j][i][w] -= until_next;
						} else {
							nsf_net->T[i][j][w] = nsf_net->T[j][i][w] = 0.0;

							/* if channel is still busy, set it free */
							if(!nsf_net->N[i][j][w]) 
								nsf_net->N[i][j][w] = nsf_net->N[j][i][w] = 1;
						}
					}
				}//for node j
			}//for node i
		}//close for gen

		fprintf(fp, "%3.4f, ", 100.0*count_block/(float)SIM_NUM_GEN);
		printf("Done\n");

		nsf_free(nsf_net);
	}//close for load

	fprintf(fp, "\n");
	fclose(fp);
	nsf_free(nsfpack);

	printf("\n");
	printf("*************************************************");
	printf(" sucess ");
	printf("*************************************************");
	printf("\n");

	return EXIT_SUCCESS;
} // close main

/*** EOF ***/
